src/OrchardCore.Modules/OrchardCore.Media/Shortcodes/ImageShortcodeProvider.cs

```diff
@@ -0,0 +1,82 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.WebUtilities;
+using Shortcodes;
+using OrchardCore.Infrastructure.Html;
+
+namespace OrchardCore.Media.Shortcodes
+{
+    public class ImageShortcodeProvider : NamedShortcodeProvider
+    {
+        private readonly IHtmlSanitizerService _htmlSanitizerService;
+        private readonly IMediaFileStore _mediaFileStore;
+
+        public ImageShortcodeProvider(IMediaFileStore mediaFileStore, IHtmlSanitizerService htmlSanitizerService)
+            : base(
+            new Dictionary<string, ShortcodeDelegate>
+            {
+                ["image"] = EvaluateImageAsync(mediaFileStore, htmlSanitizerService),
+                ["media"] = EvaluateImageAsync(mediaFileStore, htmlSanitizerService)
+            }
+        )
+        {
+            _mediaFileStore = mediaFileStore;
+            _htmlSanitizerService = htmlSanitizerService;
+        }
+
+        private static ShortcodeDelegate EvaluateImageAsync(IMediaFileStore mediaFileStore, IHtmlSanitizerService htmlSanitizerService)
+        {
+            return (args, content) =>
+            {
+                // Handle edge case of self closing shortcodes.
+                if (string.IsNullOrEmpty(content))
+                {
+                    content = args.NamedOrDefault("src");
+                    if (string.IsNullOrEmpty(content))
+                    {
+                        // Do not handle the deprecated media shortcode in this edge case.
+                        return new ValueTask<string>("[image]");
+                    }
+                }
+
+                // TODO handle virtual path, i.e. ~/
```

Comment:
@jtkech if you have a minute, I wanted to handle virtual paths here `~/` to be able to use this for static files as well. Can you point me in the right direction for how to get the virtual path, if running under tenant, or virtual folder? Thanks :)

Reply:
In the media startup we build the `DefaultMediaFileStore` with a `mediaUrlBase` combined with the `originalPathBase` and used to init a `_requestBasePath` used by `MapPathToPublicUrl()` to take into account e.g. a virtual folder

So here i would say to just skip these 2 chars `~/` and then use `MapPathToPublicUrl()`, or just skip the `~` as the path will be normalized by `MapPathToPublicUrl()`


    if (content[0] == '~')
    {
        content = mediaFileStore.MapPathToPublicUrl(content.Substring(1));
    }
    else if (!content.StartsWith("//", StringComparison.Ordinal) && !content.StartsWith("http", StringComparison.OrdinalIgnoreCase))
    {
        content = mediaFileStore.MapPathToPublicUrl(content);
    }

Reply:
@jtkech sorry, I didn't explain clearly enough.

I was hoping to support `~/` as a link to static files, that are not stored in the media file store, e.g. something like
`~/MyTheme/myembeddedbrandicon.jpg`, and also (possibly?) the files from the tenant static file provider.

Reply:
@deanmarcussen oops sorry, do you mean in the context of e.g. a `StaticFileShortcodeProvider`? Or maybe a `ResourceUrlShortcodeProvider`. As i understand it would be a scoped service and used to generate an url e.g. from `~/MyTheme/brandicon.jpg` to `{virtualFolder}/{tenantPrefix}/MyTheme/brandicon.jpg`

First, we can use the request base path including an eventual virtual folder and the tenant url prefix, but because in a given request scope in the context of a given tenant, we can create a scope and render things in the context of another tenant, normally we would need to use the original base path (before it includes any tenant url prefix) and append to it the tenant url prefix related to the current shell scope (not the request scope). That's why i made an httpContext feature holding the original base path, but maybe it's too much here ;) and this is not what we are doing most of the time.

So here i would say to just do what we do in the related liquid filters, e.g. the `ResourceUrlFilter` that uses the request pathBase through an `_httpContextAccessor` to generate an url. Then, if you want to support versioning, what we do in the `AppendVersionFilter` by using an `IFileVersionProvider`. Finally, if i understand correctly, because here you just have to generate an url, you don't have to care about any satic file provider, e.g. the tenant static file provider, they will just work as when we generate an url through liquid filters.

Reply:
awesome @jtkech I understand how that side of the system works now, and why we use the httpContent feature in the media file store, but yes I agree it is too much here. Cheers

This will be a little used feature, I suspect, but wanted to support it, so will do as per the liquid filters.

I'll hold of on the versioning feature for it, as Seb has some suggestions for refactoring that a little to avoid the file watchers, re https://github.com/OrchardCMS/OrchardCore/issues/6183

Reply:
@deanmarcussen Okay cool

Yes, i understand about file watchers, would need more time but i took a quick look in #6183 where seb said

> Can we try to implement the cache with the FileInfo modified time? This doesn't require FileWatcher

Hmm, as i understand we would have to grab the file info, at least periodically, to see maybe first if the file still exists and then if its last modified time has changed, a kind of a light file watcher. Or if it is fast enough (which is the case for embedded files) maybe grab this file info each time we generate an url, if so maybe we could just append something re-computed from this modified time (if it has changed) in place of doing a hash on the whole file. Hmm, maybe worth to do at least for embedded files so without having to compute a hash for all of them.

Reply:
cool, I've implemented for the `~/` virtual path.

If you could take a look, I'll probably merge it after you approve, before moving onto other parts of the feature for other prs.

Also I refactored to use the `IShortcodeProcessor` interface directly, to improve perf on the old `[media]` tag.

Reply:
@deanmarcussen 

Okay just did a little review otherwise LGTM